[ZHANG TONGJUN] iP

albertZhangTJ opened the pull request:
# Project Olivia

> “Your mind is for having ideas, not holding them.” – David Allen([source](https://dansilvestre.com/productivity-quotes/))

Project Olivia is a ~~CLI~~ *(GUI is on the way)* memo list software.

Currently the following commands are supported:
- `event`
- `deadline`
- `todo`
- `list`
- `ls`
- `help`
- `find`
- `mark`
- `unmark`
- `bye`
- `exit`


The difference between `event`, `deadline`, and `todo` is:
1. `event` is a task with both start time and end time
2. `deadline` is a task with a single time point
3. `todo` is a task with no time 


The service can be started with the following code:
```java
EventDispatcher worker_1=new EventDispatcher(new Calendar(), new UiHandler(), new FileHandler());
worker_1.startWorking();
```


The next step in the development process will be:
- [ ] add GUI using JavaFX :dizzy_face:

albertZhangTJ committed:
updated .gitignore to avoid uploading .class files

albertZhangTJ committed:
Level-1

albertZhangTJ committed:
Level-2

albertZhangTJ committed:
Level-3

albertZhangTJ committed:
Level-4

albertZhangTJ committed:
A-TextUiTesting

albertZhangTJ committed:
Level-5

albertZhangTJ committed:
Level-6: deleting item from calendar

albertZhangTJ committed:
Updated README

albertZhangTJ committed:
added Gradle to the existing project

albertZhangTJ committed:
Modify identifiers to match the coding standard

albertZhangTJ committed:
fix more issue with coding style

albertZhangTJ committed:
fix README logo formatting issue

albertZhangTJ committed:
fix README logo formatting issue

albertZhangTJ committed:
remove CLI logo from README

albertZhangTJ committed:
Level-7: Add syncing with disk

albertZhangTJ committed:
Convert time from String to CalendarTime

albertZhangTJ committed:
Merge pull request #1 from albertZhangTJ/branch-Level-8

Convert time from String to CalendarTime

albertZhangTJ committed:
Merge branch 'master' into branch-Level-7

albertZhangTJ committed:
Merge pull request #2 from albertZhangTJ/branch-Level-7

Level-7: Add syncing with disk

albertZhangTJ committed:
test after merging

albertZhangTJ committed:
Organize files into pacakges

albertZhangTJ committed:
Exporting as JAR

albertZhangTJ committed:
Added JUnit test

albertZhangTJ committed:
add JavaDoc comments

albertZhangTJ committed:
change things to match coding style requirements

albertZhangTJ committed:
add search functionality through find command

albertZhangTJ committed:
Merge pull request #3 from albertZhangTJ/branch-A-JavaDoc

add JavaDoc comments

albertZhangTJ committed:
Merge pull request #4 from albertZhangTJ/branch-A-CodingStandard

change things to match coding style requirements

albertZhangTJ committed:
Merge pull request #5 from albertZhangTJ/branch-Level-9

add search functionality through find command

tingkai-mai commented on src/main/java/org/Olivia/EventDispatcher.java:
```diff
@@ -0,0 +1,221 @@
+        return 200;
+    }
+
+    private int markAsDoneUndone(String input) throws Exception {
```
I appreciate the naming! Perhaps something like "toggleMarkedStatus" might be better to convey the "DoneUndone" idea?

tingkai-mai commented on src/main/java/org/Olivia/EventDispatcher.java:
```diff
@@ -0,0 +1,221 @@
+    public EventDispatcher(Calendar table, UiHandler ui, FileHandler disk) {
+        this.table=table;
+        this.ui=ui;
+        this.disk=disk;
```
There should be a space between the variables in `this.disk=disk`! I think this can be applied to the rest of your code.

tingkai-mai commented on src/main/java/org/Olivia/EventDispatcher.java:
```diff
@@ -0,0 +1,221 @@
+        return 200;
+    }
+
+    @Deprecated
```
I think if you plan to use this tag, you might want to explain what function replaces this depreciated one (https://www.baeldung.com/java-deprecated)

tingkai-mai commented on src/main/java/org/Olivia/EventDispatcher.java:
```diff
@@ -0,0 +1,221 @@
+    }
+
+    /**
+     * calls corresponding functions based on the input command
```
I think these documentation should follow the recommended Java style guide (https://se-education.org/guides/conventions/java/intermediate.html), e.g. calls should be capitalized, and sentence should end with a fullstop.

tingkai-mai commented on src/main/java/org/Olivia/EventDispatcher.java:
```diff
@@ -0,0 +1,221 @@
+     * @param input the array of string that contains the command and parameters
+     * @return status code (adapted from http, with exceptions such as status 0 represents exit)
+     */
+    public int dispatchCommand(String input) throws Exception{
```
Add space between `Exception` and `{`

tingkai-mai commented on src/main/java/org/Olivia/Olivia.java:
```diff
@@ -0,0 +1,36 @@
+    public static void main(String[] args) {
+        //each worker is attached to a calendar table to work on and an IO_handler to handle IO
+        //this design would allow the program to handle multi-user with multi-input-source case (with multi-thread)
+        //the IO_handler can be binded to a specific InputStream and a specific PrintStream (default to be stdin & stdout)
```
What does the IO_handler here refer to? I understand it's probably `UiHandler`, perhaps consider making this clearer?

albertZhangTJ replied:
Sorry, I forgot to change the occurrences in the comments. That is indeed `UiHandler`. In the previous versions it was named `IO_handler`.

tingkai-mai commented on src/main/java/org/Olivia/EventDispatcher.java:
```diff
@@ -0,0 +1,221 @@
+        return 200;
+    }
+
+    private int addEntryToCalendar(String input) throws Exception {
```
Might want to consider handing this off to a Parser class or something equivalent! Actually, this can be said for the rest of the code as I notice you often perform string parsing. Perhaps a Parser class with static helper functions may be useful in making your code more readable?

albertZhangTJ replied:
Actually the `EventDispatcher` class is designed as the parser. The input is first examined by the `dispatchCommand` function, then parsed and handled by the corresponding functions within the class.

tingkai-mai reviewed (commented):
I think this code is pretty sophisticated! It's well-written, and a 

However, I think code readability can be better improved (e.g. avoiding reliance on many if statements and replacing with switch statements, and following more closely Java coding standard: https://se-education.org/guides/conventions/java/intermediate.html). 

Perhaps you may also consider the code quality standards (https://nus-cs2103-ay2223s1.github.io/website/se-book-adapted/chapters/codeQuality.html), especially SLAP and "Make the Happy Path Prominent".

albertZhangTJ committed:
add basic GUI

jeromehjj commented on src/main/java/org/Olivia/EventDispatcher.java:
```diff
@@ -0,0 +1,230 @@
+            return list();
+        }
+
+        String[] splited_input = input.toLowerCase().split(" ");
```
Probably can consider naming it as inputs rather than splited_input

jeromehjj commented on src/main/java/org/Olivia/EventDispatcher.java:
```diff
@@ -0,0 +1,230 @@
+        catch (Exception e) {
+            ui.cout(e.getMessage() + "\n");
+            ui.cout("Error syncing with disk, starting with a fresh new table" + "\n");
+        }
```
The try-catch statement should have the follow form: 
```suggestion
        try {
            this.disk.syncFromFile(this.table);
        } catch (Exception e) {
            ui.cout(e.getMessage() + "\n");
            ui.cout("Error syncing with disk, starting with a fresh new table" + "\n");
        }
```

jeromehjj commented on src/main/java/org/Olivia/EventDispatcher.java:
```diff
@@ -0,0 +1,230 @@
+                    break;
+                }
+            }
+            catch (Throwable e) {
```
Same as mentioned above.

jeromehjj commented on src/main/java/org/Olivia/GuiEventDispatcher.java:
```diff
@@ -0,0 +1,177 @@
+package org.Olivia;
+/**
+ * Listens command from a given UiHandler
```
I think this could be placed further down right above the class block.

jeromehjj commented on src/main/java/org/Olivia/GuiEventDispatcher.java:
```diff
@@ -0,0 +1,177 @@
+            }
+
+        }
+        else if (args[0].equals("event") && args.length >= 4) {
```
I think it should be as following:
```suggestion
        } else if (args[0].equals("event") && args.length >= 4) {
```

jeromehjj commented on src/main/java/org/Olivia/IO/FileHandler.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    private CalendarEntry parseEntry(String line) throws Exception {
+        //System.out.println("Parsing "+line);
```
Perhaps you might want to remove this if this is unused?

jeromehjj commented on src/main/java/org/Olivia/IO/FileHandler.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    public int syncFromFile(Calendar c) throws Exception {
+        File fd = new File(this.fileName);
+        if (!fd.exists()) {
+            return 200;
+        }
+        Scanner file = new Scanner(fd);
```
I think it might be better to explicitly name the File object and Scanner object as file and scanner respectively to make things clearer for reading 😅

jeromehjj commented on src/main/java/org/Olivia/Launcher.java:
```diff
@@ -0,0 +1,32 @@
+ */
+public class Launcher {
+    public static void main(String[] args) {
+//        try {
```
I think this can be remove too if unused.

jeromehjj commented on src/main/java/org/Olivia/Olivia.java:
```diff
@@ -0,0 +1,163 @@
+        catch (Exception e) {
+            e.printStackTrace();
+        }
+        this.dispatcher=new GuiEventDispatcher(new Calendar(), new FileHandler());
```
I think you forgot to add spacing surrounding the operator here.

jeromehjj reviewed (commented):
LGTM! I think that your code looks good and most of it complies with the coding standards. Just need to make those changes highlighted and it should be good to go! Good job! 👍

albertZhangTJ committed:
fixed several issues with the GUI

albertZhangTJ committed:
Merge pull request #6 from albertZhangTJ/branch-Level-10

fixed several issues with the GUI

albertZhangTJ committed:
Build a uber jar with gradle

Updated project Gradle to version 7.2. Installed Shadow plugin for gradle version 7.1.2.

Exported project as a uber jar.

albertZhangTJ committed:
Fix issue with locating icon file

Using canonical path in getResourceAsStream

albertZhangTJ committed:
Merge pull request #7 from albertZhangTJ/branch-A-Assertions

Fix issue with locating icon file

albertZhangTJ committed:
Use assertions to check workflow

Use Java `assert` to check whether componenets with dependencies are initialized in a proper order.

albertZhangTJ committed:
Merge pull request #8 from albertZhangTJ/branch-A-Assertions

Use assertions to check workflow

albertZhangTJ committed:
refactor FileHandler

albertZhangTJ committed:
Merge pull request #9 from albertZhangTJ/branch-A-CodeQuality

refactor FileHandler

albertZhangTJ committed:
Add support for tagging tasks

Users are able to add `#<tag content>` at end of the command to tag a task

albertZhangTJ committed:
update help message

albertZhangTJ committed:
added UG

albertZhangTJ committed:
added UG

albertZhangTJ committed:
Update GUI

albertZhangTJ committed:
Update Ui.png

albertZhangTJ committed:
Fix bug with tags; refactored the code

albertZhangTJ committed:
Add comments

albertZhangTJ committed:
Add comments

albertZhangTJ committed:
Add comments

albertZhangTJ committed:
Improve code style
